### PR TITLE
Feature champion level

### PR DESCRIPTION
**Pull request by PaulQbFeng**
New feature: Champions can now be initialised with a level from 1 to 18. base stats are updated accordingly

### CODE REVIEW

**Comment by PaulQbFeng** [comment]
Btw should I add `level` as an argument again to be more explicit in the `update_stat_from_level()` method ?

**Comment by PaulQbFeng** [comment]
>The growth coefficient is based on the average growth per level between level 1 and level 18. The actual base attack damage does not scale linearly.

https://leagueoflegends.fandom.com/wiki/List_of_champions/Attack_damage
/ff

**Comment by PaulQbFeng** [comment]
But maybe ????? 
image

**Comment by PaulQbFeng** [comment]
https://leagueoflegends.fandom.com/wiki/Champion_statistic

File: champion.py [hunk 1]
```diff
@@ -1,20 +1,33 @@
...
 def fill_champion_stats(dataset: dict):
     # define champion list
     champion_list = list(dataset["data"].keys())
-    ALL_CHAMPION_BASE_STAT={}
+    ALL_CHAMPION_BASE_STAT = {}
     for x in champion_list:
-        ALL_CHAMPION_BASE_STAT[x]=dataset["data"][x]["stats"]
-    return(ALL_CHAMPION_BASE_STAT)
+        ALL_CHAMPION_BASE_STAT[x] = dataset["data"][x]["stats"]
+    return ALL_CHAMPION_BASE_STAT
```
Comment (Lysxia) [inline champion.py:17]: Dictionary comprehension FTW
```
def fill_champion_stats(dataset: dict):
    return { name: data["stats"] for name, data in dataset["data"].items() }
```

File: champion.py [hunk 2]
```diff
@@ -30,43 +43,56 @@ class BaseChampion:
         ...
     """
 
-    def __init__(self, champion_name: str):
-        self.initialize_base_stat(champion_name)
+    def __init__(self, champion_name: str, level: int):
+        assert level in np.arange(1, 19), "Champion level should be in the [1,18] range"
```
Comment (Lysxia) [inline champion.py:47]: why no
```suggestion
        assert 1 <= level <= 18
```

File: champion.py [hunk 3]
```diff
@@ -30,43 +43,61 @@ class BaseChampion:
...
+        self.base_stats = ALL_CHAMPION_BASE_STAT[champion_name]
+        self.update_stat_from_level()
 
-    def initialize_base_stat(self, champion_name):
+    def update_stat_from_level(self):
         """Takes all the base stats from the input dictionary and create the corresponding attributes in the instance"""
-        self.__dict__.update(ALL_CHAMPION_BASE_STAT[champion_name])
+
+        def calculate_flat_stat_from_level(base: float, mean_growth_perlevel: float, level: int):
+            # Check Readme for formula detail
```
Comment (Lysxia) [inline champion.py:56]: Why not put links to all references we use in comments https://leagueoflegends.fandom.com/wiki/Champion_statistic#Growth_statistic_calculations and (below) https://leagueoflegends.fandom.com/wiki/Attack_speed#Bonus_attack_speed_per_level

File: champion.py [hunk 4]
```diff
@@ -30,43 +43,61 @@ class BaseChampion:
...
 
-
 # Each champion has its own class as their spells have different effects.
 class Annie(BaseChampion):
     champion_name = "Annie"
 
-    def __init__(self):
-        super().__init__(champion_name=__class__.champion_name)
+    def __init__(self, level: int = 1):
+        super().__init__(champion_name=__class__.champion_name, level=level)
```
Comment (Lysxia) [inline champion.py:82]: Would this be better?
```suggestion
    def __init__(self, **kwargs):
        super().__init__(champion_name=__class__.champion_name, **kwargs)
```
and move the default argument to `BaseChampion`. And when `BaseChampion` gets more default arguments, you won't need to change each champion's class.

File: champion.py (hunk 2 above)
Comment (PaulQbFeng) [reply champion.py:47]: I wanted to limit the level to an **integer** between 1 and 18. what else do you suggest ? isnstance(level,int) and  1<= level <= 18 ?

Comment (Lysxia) [reply champion.py:47]: doesn't the annotation `level: int` take care of that?

Comment (PaulQbFeng) [reply champion.py:47]: In python we use a type **hinting** system ... ^^'

**Pull request merged**